rng/src/main.rs

```diff
@@ -0,0 +1,170 @@
+            return Err(Error::InvalidPeriodInput);
+        }
+
+        if args.socket_count == 0 {
+            return Err(Error::InvalidSocketCount);
+        }
+
+        let socket_path = match args.socket_path {
+            Some(path) => path,
+            None => return Err(Error::InvalidSocketPath),
+        };
+
+        // Divide available bandwidth by the number of threads in order
+        // to avoid overwhelming the HW.
+        let max_bytes = args.max_bytes / args.socket_count as usize;
+
+        Ok(VuRngConfig {
+            period_ms: args.period,
+            max_bytes,
+            count: args.socket_count,
+            socket_path,
+            rng_source: args.rng_source,
+        })
+    }
+}
+
+pub fn start_backend(config: VuRngConfig) -> Result<()> {
+    let mut handles = Vec::new();
+    let random_file = Arc::new(Mutex::new(
+        File::open(&config.rng_source).map_err(|_| Error::AccessRngSourceFile)?,
+    ));
+
+    for i in 0..config.count {
+        let socket_path = config.socket_path.clone();
+        let socket = format!("{}{}", socket_path, i);
+        let period_ms = config.period_ms;
+        let max_bytes = config.max_bytes;
+        let random = Arc::clone(&random_file);
+
+        let handle = thread::spawn(move || loop {
+            let vu_rng_backend = Arc::new(RwLock::new(
+                VuRngBackend::new(random.clone(), period_ms, max_bytes).unwrap(),
+            ));
+
+            let mut daemon = VhostUserDaemon::new(
+                String::from("vhost-user-RNG-daemon"),
+                vu_rng_backend.clone(),
+                GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+            )
+            .unwrap();
+
+            let listener = Listener::new(socket.clone(), true).unwrap();
+            daemon.start(listener).unwrap();
+
+            match daemon.wait() {
+                Ok(()) => {
+                    info!("Stopping cleanly.");
+                }
+                Err(vhost_user_backend::Error::HandleRequest(
+                    vhost_user::Error::PartialMessage,
+                )) => {
+                    info!("vhost-user connection closed with partial message. If the VM is shutting down, this is expected behavior; otherwise, it might be a bug.");
+                }
+                Err(e) => {
+                    warn!("Error running daemon: {:?}", e);
+                }
+            }
+
+            // No matter the result, we need to shut down the worker thread.
+            vu_rng_backend
+                .read()
+                .unwrap()
+                .exit_event
+                .write(1)
+                .expect("Shutting down worker thread");
+        });
+
+        handles.push(handle);
+    }
+
+    for handle in handles {
+        handle.join().map_err(|_| Error::FailedJoiningThreads)?;
+    }
+
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    env_logger::init();
+
+    start_backend(VuRngConfig::try_from(RngArgs::parse()).unwrap())
```

Comment:
I would rather no move forward with this one because:

1. From an architectural perspective parsing command line arguments and starting the backend are two different things that should be decoupled.  

2. On the RNG side it decreases the test coverage because VuRngConfig is not cloned, something that is currently done in verify_start_backend().